articles/embedded.md

```diff
@@ -0,0 +1,309 @@
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+While standard ROS 2 can run on Linux-class embedded devices, smaller 
+microcontrollers (MCU) are still a challenge. For example, ARM Cortex-M3 
+and -M4 class MCUs are popular for motion control and sensor integration, 
+but have little RAM and storage. ARM Cortex-M3 is widely used in many 
+sensors due to its optimized power consumption capabilities and ultra low 
+power modes. These microcontrollers are often running with small batteries, 
+e.g. for environmental sensors or small consumer robots, which requires
+efficient power management. We aim to address these devices through a 
+combination of specialized, ROS 2-interoperable stack as well as the use
+of Real-Time Operating Systems (RTOS).
+
+This article documents feature wishes, requirements, design ideas and 
+related work on this undertaking.
+
+## Wishlist
+
+A number of things that would be great to have, lets call it a wishlist. 
+Feel free to add stuff ;-)
+
+### W-SEAMLESS: *Seamless integration with ROS&nbsp;2*
+
+The embedded ROS 2 stack shall integrate seamlessly with standard 
+ROS 2 stacks and nodes running on more powerful microprocessors. For 
+a standard ROS 2 node, communication (topics, services) from and to
+software running on the embedded ROS 2 stack should be transparent. 
+The same should hold for other core concepts such as parameters, 
+graph introspection, and run-time reconfiguration by the node lifecycle.
+
+### W-PORTABILITY: *Portability of ROS 2-based software*
+
+The embedded ROS 2 stack shall resemble or directly use the ROS 2 API 
+-- more precisely the rclcpp API -- to facilitate porting standard 
+ROS 2 nodes to MCUs.
+
+### W-DEVICES *Support of a broad spectrum of device classes*
+
+The embedded ROS 2 stack shall support a broad range of MCU device 
+classes, starting from a few tens kilobyte of RAM.
+
+### W-RTOS: *Support of a broad spectrum of RTOS(s)* 
+
+The stack shall be adaptable to different RTOS (e.g., NuttX, FreeRTOS, 
+Zephyr) and provide abstractions with regard to the application 
+software to facilitate porting application components to between RTOSs.
+
+### W-COMM: *Support of prevalent communication technologies* 
+
+To cover the broad range of use-cases for MCUs in robotics, the embedded 
+ROS 2 stack shall be usable with the default ROS 2 middleware standard 
+DDS, simple (custom) serial communication protocols just as common 
+wireless technologies like 6LoWPAN
+
+### W-MODULARITY: *High Modularity* 
+
+The embedded ROS 2 stack shall be highly modular, not only to allow for 
+adaptation to specific hardware features and different RTOS, but also to
+allow the integration with existing frameworks and to be able to create 
+customized sub-stacks. For example, it shall be usable in the style of 
+rosserial, thus bringing basically the message serialization to the MCU 
```

Comment:
If there are use cases that are only satisfiable using a rosserial-style approach, then I think a better approach is to create a rosserial2 for those and improve the existing ROS2 stack for the rest.